stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1385 @@
+                    &parent_block_id
+                )?.ok_or_else(|| {
+                    warn!("While processing tenure change, failed to look up parent block's total tx fees";
+                          "parent_block_id" => %parent_block_id,
+                          "block_hash" => %block_hash,
+                          "block_consensus_hash" => %chain_tip_consensus_hash);
+                    ChainstateError::NoSuchBlockError
+                })?
+            } else {
+                // if the parent tenure is an epoch-2 block, don't pay
+                // any fees to them in this schedule: nakamoto blocks
+                // cannot confirm microblock transactions, and
+                // anchored transactions are scheduled
+                // by the parent in epoch-2.
+                0
+            };
+
+            Some(
+                Self::make_scheduled_miner_reward(
+                    mainnet,
+                    evaluated_epoch,
+                    &parent_tenure_header.anchored_header.block_hash(),
+                    &parent_tenure_header.consensus_hash,
+                    &block_hash,
+                    block
+                        .get_coinbase_tx()
+                        .ok_or(ChainstateError::InvalidStacksBlock(
+                            "No coinbase transaction in tenure changing block".into(),
+                        ))?,
+                    chain_tip_consensus_hash,
+                    next_block_height,
+                    parent_tenure_fees,
+                    burnchain_commit_burn,
+                    burnchain_sortition_burn,
+                    total_coinbase,
+                )
+                .expect("FATAL: parsed and processed a block without a coinbase"),
+            )
+        } else {
+            None
+        };
+
+        let matured_rewards_info = miner_payouts_opt
+            .as_ref()
+            .map(|(_, _, _, info)| info.clone());
+
+        let new_tip = Self::advance_tip(
+            &mut chainstate_tx.tx,
+            &parent_chain_tip.anchored_header,
+            &parent_chain_tip.consensus_hash,
+            &block.header,
+            chain_tip_consensus_hash,
+            chain_tip_burn_header_hash,
+            chain_tip_burn_header_height,
+            chain_tip_burn_header_timestamp,
+            scheduled_miner_reward.as_ref(),
+            miner_payouts_opt,
+            &block_execution_cost,
+            &total_tenure_cost,
+            block_size,
+            applied_epoch_transition,
+            burn_stack_stx_ops,
+            burn_transfer_stx_ops,
+            burn_delegate_stx_ops,
+            affirmation_weight,
+            tenure_height,
+            tenure_changed,
+            block_fees,
+        )
+        .expect("FATAL: failed to advance chain tip");
+
+        chainstate_tx.log_transactions_processed(&new_tip.index_block_hash(), &tx_receipts);
+
+        monitoring::set_last_block_transaction_count(block.txs.len() as u64);
```

Comment:
Does `usize` always convert to `u64`?

Reply:
`usize` is based on pointer size. If there's some platform with 128-bit addressing (I doubt it), I guess `u64::from()` would fail to compile on it

I know these `as` conversions probably won't ever truncate in real world use, but these sorts of silent errors are a real pain to debug and it's trivial to completely avoid them, so it's better just not to `as` at all